db/system_distributed_keyspace.cc

```diff
@@ -152,8 +155,14 @@ static const sstring CDC_TIMESTAMPS_KEY = "timestamps";
 schema_ptr service_levels() {
     static thread_local auto schema = [] {
         auto id = generate_legacy_id(system_distributed_keyspace::NAME, system_distributed_keyspace::SERVICE_LEVELS);
-        return schema_builder(system_distributed_keyspace::NAME, system_distributed_keyspace::SERVICE_LEVELS, std::make_optional(id))
+        auto builder = schema_builder(system_distributed_keyspace::NAME, system_distributed_keyspace::SERVICE_LEVELS, std::make_optional(id))
                 .with_column("service_level", utf8_type, column_kind::partition_key)
+                .with_column("shares", int32_type);
+        if (utils::get_local_injector().is_enabled("service_levels_v1_table_without_shares")) {
```

Comment:
Yes, the purpose of this error injection is to simulate an upgrade from a version which didn't support workload prioritization. It was introduced in scylladb/scylla-enterprise#4634.

For the record, we do have a proper upgrade test in dtests which does not do any error injections (enterprise_upgrade_test.py::TestEnterpriseUpgradeFromOSS).

Reply:
So maybe we don't need this injection thing at all. These injections all over the code are really starting to become unsightly (and too frequently - hard to understand, and we'll really hate them if one day we want to refactor code containing injections). But I guess that if you're just copying code that already existed in Enterprise and had this injection, we don't have to rethink this old decision right now :-(

Reply:
I'd rather keep it for now because it does some checks that the dtest doesn't. I'm not a great fan of the error injection in this particular place either, I think it was mostly done for convenience as ./test.py doesn't support proper rolling upgrade tests (yet?). It would be great to unify those tests, make them do a "real" rolling upgrade and put them into one place, but that would be a larger effort than I'd like to do in this PR.

Reply:
@nyh you are reviewing existing code that is being ported. Any changes should be applied after the merge, or we will diverge.